src/librustc_mir/borrow_check/mod.rs

```diff
@@ -316,7 +317,12 @@ fn do_mir_borrowck<'a, 'gcx, 'tcx>(
             }
 
             let span = local_decl.source_info.span;
-            let mut_span = tcx.sess.source_map().span_until_non_whitespace(span);
+            let source_map = tcx.sess.source_map();
+            let mut_span = source_map.span_until_non_whitespace(span);
+            if !source_map.span_to_string(mut_span).contains(&*keywords::Mut.name().as_str()) {
+                // If the `mut` appears in desugared code, we ought not to warn about it.
+                continue;
+            }
```

Comment:
Hmm, re-reading this, it seems kind of strange. I didn't notice at first. Searching the span for the `mut` keyword feels hacky, though — if the problem is that this variable is introduced in a HIR desugaring step, we ought to be able to check the "expansion trace" for the span to see if it arises from a desugaring, no?

Reply:
Yeah, I was planning to rewrite this bit; it feels messy.

Reply:
Okay, I've fixed this.